src/bigistore.c

```diff
@@ -1077,28 +1077,35 @@ Datum bigistore_slice_min_max(PG_FUNCTION_ARGS)
         ereport(ERROR, (errcode(ERRCODE_SYNTAX_ERROR), errmsg("min must be less or equal max")));
 
     if (is->len <= 0 || pairs[is->len - 1].key < min || pairs[0].key > max)
-        PG_RETURN_NULL();
+        PG_RETURN_EMPTY_ISTORE();
 
     if (pairs[0].key >= min && pairs[is->len - 1].key <= max)
         PG_RETURN_POINTER(is);
 
+    is->buflen = 0;
+    is->len    = 0;
+
     for (i = 0; i < len; i++)
     {
-        if (pairs[i].key < min)
```

Comment:
That loop interior seems to be too complicated. I'd skip forward the first elements less than min and then process the rest lower than max (we know there is at least one, otherwise the condition at line 1079 would return early.

how about something like this (for bigistore):
```
diff --git a/src/bigistore.c b/src/bigistore.c
index e43c79e..dfb29f5 100644
--- a/src/bigistore.c
+++ b/src/bigistore.c
@@ -1085,24 +1085,15 @@ Datum bigistore_slice_min_max(PG_FUNCTION_ARGS)
     is->buflen = 0;
     is->len    = 0;

-    for (i = 0; i < len; i++)
+    while (pairs[i].key < min && ++i); // skip pairs lower than min
+    min_idx = i;
+    for (; pairs[i].key <= max && i < len; i++)
     {
-        // set the new starting index of pairs
-        if (pairs[i].key >= min && min_idx == 0)
-            min_idx = i;
-
-        if (pairs[i].key > max)
-            break;
-
-        if (pairs[i].key >= min && pairs[i].key <= max)
-        {
-            ++is->len;
-            is->buflen += bigis_pair_buf_len(pairs + i);
-        }
+        ++is->len;
+        is->buflen += bigis_pair_buf_len(pairs + i);
     }

-    if (is->len == 0)
-        PG_RETURN_EMPTY_ISTORE();
+    Assert(is->len > 0);

     if (min_idx > 0)
         memmove(pairs, pairs + min_idx, (is->len * sizeof(BigIStorePair)));
```

Reply:
See my comment below